src/optimizer/remove_unused_columns.cpp

```diff
@@ -461,14 +688,80 @@ void BaseColumnPruner::AddBinding(BoundColumnRefExpression &col, ColumnIndex chi
 		ReferencedColumn column;
 		column.bindings.push_back(col);
 		column.child_columns.push_back(std::move(child_column));
-		column_references.insert(make_pair(col.binding, std::move(column)));
+		entry = column_references.emplace(make_pair(col.binding, std::move(column))).first;
 	} else {
 		// column reference already exists - check add the binding
 		auto &column = entry->second;
 		column.bindings.push_back(col);
 
 		MergeChildColumns(column.child_columns, child_column);
 	}
+	if (mode == BaseColumnPrunerMode::DISABLE_PUSHDOWN_EXTRACT) {
+		//! Any child referenced after this mode is set disables PUSHDOWN_EXTRACT
+		entry->second.supports_pushdown_extract = false;
+	}
+}
+
+void ReferencedColumn::AddPath(const vector<idx_t> &path) {
+	if (child_columns.empty()) {
+		//! Full field already referenced, won't use struct field projection pushdown at all
+		return;
+	}
+	D_ASSERT(!path.empty());
+
+	//! Do not add the path if it is a child of an existing path
+	auto end_iterator = ++path.begin();
+	while (true) {
+		//! Create a subset of the path up to an increasing depth, so we can check if the parent path already exists
+		vector<idx_t> path_copy(path.begin(), end_iterator);
```

Comment:
same comment as above, we can add to this vector instead of rewriting it every iteration